File: algorithms/1_easy/F/01779_find_nearest_point_that_has_the_same_x_or_y_coordinate/solution.c
#define INF INT_MAX

static inline int abs_int(const int a) {
    return a < 0 ? -a : a;
}

int nearestValidPoint(const int x, const int y, int** points, const int rows, const int *restrict cols_data) {
    assert(rows > 0 && cols_data != NULL && points != NULL);
    assert(x >= 1 && x <= 10000 && y >= 1 && y <= 10000);
    int min_dis = INF;
    int min_dis_idx = -1;

    for (int i = 0; i < rows; i += 1) {
        assert(cols_data[i] == 2);
        const int px = points[i][0];
        const int py = points[i][1];
        assert(px >= 1 && px <= 10000 && py >= 1 && py <= 10000);
        if (px == x || py == y) {
            const int curr_dis = abs_int(x - px) + abs_int(y - py);
            if (curr_dis < min_dis) {
                min_dis = curr_dis;
                min_dis_idx = i;
            }
        }
    }

    return min_dis_idx;
}

File: algorithms/1_easy/F/01779_find_nearest_point_that_has_the_same_x_or_y_coordinate/solution.rs
impl Solution {
    const INF: i32 = i32::MAX;

    pub fn nearest_valid_point(x: i32, y: i32, points: Vec<Vec<i32>>) -> i32 {
        debug_assert!(x >= 1 && x <= 10_000 && y >= 1 && y <= 10_000);
        let mut min_dis = Self::INF;
        let mut min_dis_idx = -1_isize;

        for (i, p) in points.iter().enumerate() {
            debug_assert!(p.len() == 2);
            let (px, py) = (p[0], p[1]);
            debug_assert!(px >= 1 && px <= 10_000 && py >= 1 && py <= 10_000);
            if px == x || py == y {
                let curr_dis = (x - px).abs() + (y - py).abs();
                if curr_dis < min_dis {
                    min_dis = curr_dis;
                    min_dis_idx = i as isize;
                }
            }
        }

        min_dis_idx.try_into().unwrap()
    }
}
